### DISCUSSION TITLE
Potential bug, .bashrc not read on startup of brand new user with nothing configured

### DISCUSSION THREAD

**Discussion by velicanu** [ghostty-org/ghostty#3437] [Feature Requests, Ideas] [answered]
Hello, brand new user to Ghostty excited to try it out. I ran into a problem when first launching ghostty so I thought I'd share since it may impact other new users, of .bashrc not being picked up at first launch. Here are some more details:

OS: MacOS 15.1 (24B2082) (M4)
Install method: brew install ghostty
```
ghostty --version
Ghostty 1.0.0

Version
  - version: 1.0.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
```
coming from iTerm2

I have absolutely nothing configured in ghostty, just my .bashrc file  located at `~/.bashrc` and no `~/.bash_profile` . The first few lines of my bashrc file look like this, with more stuff below:
```
# .bashrc
export HISTSIZE=100000
export HISTFILESIZE=100000
export HISTCONTROL=ignoreboth:erasedups
shopt -s histappend
PROMPT_COMMAND="history -a;$PROMPT_COMMAND"
```

An unfortunate byproduct of not reading the .bashrc is that my `bash_history` has been truncated to 500 lines the first time I ran ghostty 😞  , might be good to look into this so other new users don't run into a similar pickle.

**Comment by velicanu** [comment]
forgot to mention, running `source ~/.bashrc` after the terminal loads from ghostty works.

**Comment by mitchellh** [comment]
Hm. Cc @jparise

**Comment by jparise** [comment]
I'm not able to reproduce this yet, but I'll continue to look for the root cause.

@velicanu, can you say which version of bash you're using so I can continue to narrow down your environment? In particular, it would be helpful to know if this is the default macOS `/bin/bash` (version 3.2) or another version that you've installed via e.g. homebrew.

**Comment by velicanu** [comment]
This is homebrew bash:
```
m4 ~ $ which bash
/opt/homebrew/bin/bash
m4 ~ $ echo $SHELL
/opt/homebrew/bin/bash
m4 ~ $ bash --version
GNU bash, version 5.2.37(1)-release (aarch64-apple-darwin24.0.0)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by jparise** [comment]
And I assume setting `shell-integration = none` fixes it?

Also, what permissions do you have set on your `.bashrc` file?

**Comment by velicanu** [comment]
> And I assume setting `shell-integration = none` fixes it?

It doesn't seem to fix it, I put that line in the config and restarted ghostty but the behavior is the same.

> Also, what permissions do you have set on your `.bashrc` file?

```
m4 ~ $ ls -tralp .bashrc
-rw-r--r--@ 1 dragosvelicanu  staff  9329 Dec 26 15:53 .bashrc
```

seems to be default perms that new files get

```
m4 ~ $ touch mynewfile
m4 ~ $ ls -tralp mynewfile
-rw-r--r--@ 1 dragosvelicanu  staff  0 Dec 27 10:05 mynewfile
```

**Comment by mitchellh** [comment]
Okay, moving this to issue triage since I thought this would be easy to confirm. I can't reproduce this either but I'm also not a heavy bash user. Once we get more accurate details we can convert this to an issue again.

**Reply by velicanu** [reply]
Sounds good, let me know if there's anything I can do or run on my end to help out or provide more info.

**Reply by jparise** [reply]
It might be helpful to see more of your `.bashrc` in case there are some early conditions or error cases that we're hitting.

**Reply by velicanu** [reply]
I've attached my full bashrc, mostly unedited except for some sensitive env vars removed.

[bashrc.txt](https://github.com/user-attachments/files/18261616/bashrc.txt)

**Reply by velicanu** [reply]
I suspect the root cause is not related to the contents of the .bashrc file. I've truncated my .bashrc to just 3 lines and it's still not being read

<img width="472" alt="Screenshot 2024-12-27 at 12 25 14 PM" src="https://github.com/user-attachments/assets/f2714d56-f7d7-4602-90d7-0d22fd5e6289" />

**Reply by jparise** [reply]
I also don't see anything in your `.bashrc` that would cause this kind or problem. And given that `shell-integration = none` doesn't solve it, that also rules out our shell integration code as a source of the problem.

I can only think of two other things:

1. It has something to do with the way we're executing `bash` as the login shell, and that's bumping into something on your system that others haven't seen before.
2. There's something about your system-level `/etc/bashrc` that's causing problems within Ghostty. Maybe you have a `/etc/bashrc_$TERM_PROGRAM` file (which is something `/etc/bashrc` on macOS will source)?

**Reply by velicanu** [reply]
I did a little more digging on my end and it seems that ghostty looks for a `.bash_profile` file and iTerm2 (my previous terminal emulator) looks for a `.bashrc` file. Since:

1. if I have a `.bashrc` file and have no `.bash_profile` file, ghostty reads nothing and iTerm2 loads properly whenever I start a new terminal
2. if i have a `.bash_profile` file and have no `.bashrc` file, ghostly loads properly and iTerm2 reads nothing whenever I start a new terminal

I have no idea if this is expected behavior or what the standards are around which file gets read, but I'm happy to mark this as resolved from my end since I can work around it and try out ghostty by making the `.bash_profile` file. Also happy to help with further debugging if you guys want to pursue.

**Reply by jparise** [reply, marked as answer]
You can read more about bash's startup files here: https://www.gnu.org/software/bash/manual/bash.html#Bash-Startup-Files

It's common to source your `.bashrc` from `.bash_profile` to avoid some of the confusing things you've encountered:

```bash
# .bash_profile
[ -r ~/.bashrc ] && . ~/.bashrc
```

And also to avoid doing work in `.bashrc` when your shell is not interactive:

```bash
# .bashrc
[ -z "$PS1" ] && return
```

**Comment by cngeeks** [comment]
Mine situation is not the same. ghostty will source .bashrc but not .bash_profile by default even i have code to source .bashrc in my .bash_profile.

then if i add sourcing .bash_profile in .bashrc, then ghostty will source both correctly but this is against the bash startup policy and breaks other terminal behavior.

I'm running deepin 23 and all permissions of the profiles are correct.

❯ ghostty --version
Ghostty 1.0.1

Version
  - version: 1.0.1
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.6.9
    runtime    : 4.12.3
  - libadwaita : enabled
    build      : 1.1.7
    runtime    : 1.5.0
  - libX11     : enabled

**Reply by jparise** [reply]
Our automatic shell integration should be sourcing your `.bash_profile` under these conditions:

1. bash is invoked as a login shell
2. the bash command line doesn't include the `--noprofile` option
3. the `~/.bash_profile` file is readable

https://github.com/ghostty-org/ghostty/blob/918ccdba5cc65ccd1fb48a54c71306d869299441/src/shell-integration/bash/ghostty.bash#L47-L53

**Reply by cngeeks** [reply]
i click on the ghostty desktop shortcut to bring up ghostty. does this mean bash is invoked as a login shell?
and i did not add --noprofile anywhere.
~/.bash_profile is readable and it's working fine with other terminals.

plus, i do have installed oh-my-bash which is working fine with other terminals.

**Reply by jparise** [reply]
From the [bash manual](https://www.gnu.org/software/bash/manual/html_node/Invoking-Bash.html):

> A login shell is one whose first character of argument zero is ‘-’, or one invoked with the --login option.

You can check if your session is a login shell using:

```sh
$ shopt login_shell
login_shell     on
```

On macOS, we force a login shell for "reasons":

https://github.com/ghostty-org/ghostty/blob/918ccdba5cc65ccd1fb48a54c71306d869299441/src/termio/Exec.zig#L951-L962

... but I see you're in a Linux environment, so the expectation is that terminal sessions are _not_ login sessions by default (i.e. just running `bash` is not a login shell). Instead, if you want login shell behavior, you'll need to use `bash --login`. You can do that by changing Ghostty's `command` configuration option (e.g. `command = bash --login`). I think your other terminals might be doing that, but I'm not a Linux GUI user, so others might have better first-hand knowledge to share here.

**Reply by cngeeks** [reply]
@jparise i'm using both linux and macos and my issue only exist on linux.
and you are right about `command` option.
it works fine after i add `initial-command = bash --login` into `~/.config/ghostty/config`
thxxx

**Reply by NDex30** [reply]
I have a similar issue, on my Mac OS locally ghosty works fine, however when I remote in to a virtual ubuntu workstation thats on bash, it doesn't seem to process the `.bashrc` or the `.bash_profile`. I was wondering if there is something that needs installed on the remote machine or if both need to be on zsh?

**Reply by jparise** [reply]
@NDex30 launching the remote (Ubuntu) shell session is out of Ghostty's direct control; it's started by sshd on that system.

**Reply by NDex30** [reply]
Ahh I apologize as I mispoke after doing more testing, it looks like most of the setup was completed, but the PS1 changes aren't picked up. iTerm2 shows the configured PS1 when I remote in, but ghostty does not. Perhaps I just have something misconfigured.
